File: backend/app/schemas/monitor.py
"""
@Author: li
@Email: lijianqiao2906@live.com
@FileName: monitor.py
@DateTime: 2026-08-13 14:00
@Docs: 监控目标请求与响应模型
"""

from datetime import datetime
from typing import Literal, Self

from pydantic import ConfigDict, Field, model_validator

from app.schemas.common import ApiModel

type MonitorLatestStatus = Literal["up", "down"]


class MonitorTargetCreate(ApiModel):
    """创建监控目标。"""

    ip_address: str = Field(min_length=1, max_length=45)
    port: int = Field(ge=1, le=65535)
    label: str = Field(default="", max_length=100)
    check_interval_seconds: int = Field(default=30, ge=5, le=3600)
    is_active: bool = True
    cmdb_asset_id: int | None = Field(default=None, gt=0)


class MonitorTargetUpdate(ApiModel):
    """部分更新监控目标；未出现的字段保持原值。"""

    ip_address: str | None = Field(default=None, min_length=1, max_length=45)
    port: int | None = Field(default=None, ge=1, le=65535)
    label: str | None = Field(default=None, max_length=100)
    check_interval_seconds: int | None = Field(default=None, ge=5, le=3600)
    is_active: bool | None = None
    cmdb_asset_id: int | None = Field(default=None, gt=0)

    @model_validator(mode="after")
    def reject_empty_update(self) -> Self:
        """至少提供一个要更新的字段。"""
        if not self.model_fields_set:
            raise ValueError("至少提供一个要更新的字段")
        return self


type MonitorBucketState = Literal["up", "down", "unknown"]


class MonitorUptimeWindow(ApiModel):
    """最近一小时的可用率状态条：60 个分钟格 + 可用率。

    三个字段绑在一起才有意义，所以做成嵌套对象而不是摊平：
    单看 `buckets` 不知道它从哪个时刻开始、每格多长。前端靠
    `started_at + index * bucket_seconds` 算出每格对应的时间做 tooltip。
    """

    started_at: datetime
    bucket_seconds: int
    buckets: list[MonitorBucketState]
    # 窗口内一次探测都没有时为 None，而不是 1.0——
    # 一个刚建好、从没跑过的目标显示「100% 可用」是撒谎。
    uptime_rate: float | None


class MonitorTargetResponse(ApiModel):
    """监控目标公开表示，附带最近一次探测结果与最近一小时的可用率状态条。"""

    id: int
    cmdb_asset_id: int | None
    ip_address: str
    port: int
    label: str
    check_interval_seconds: int
    is_active: bool
    created_at: datetime
    latest_status: MonitorLatestStatus | None = None
    latest_latency_ms: int | None = None
    latest_detail: str = ""
    latest_checked_at: datetime | None = None
    uptime_window: MonitorUptimeWindow

    model_config = ConfigDict(from_attributes=True)


class MonitorRuntimeResponse(ApiModel):
    """监控页轮询所需的运行参数。"""

    sweep_interval_seconds: int


class MonitorLogItem(ApiModel):
    """单条监控状态变化日志。"""

    id: int
    target_id: int
    label: str
    ip_address: str
    port: int
    status: MonitorLatestStatus
    latency_ms: int | None = None
    detail: str = ""
    checked_at: datetime

File: frontend/src/types/monitor.ts
/** 监控目标相关类型 */

export type MonitorLatestStatus = "up" | "down"

/** 监控页轮询间隔（来自系统配置中的全局巡检间隔） */
export interface MonitorRuntime {
  sweep_interval_seconds: number
}

/** 状态条里一格的状态。`unknown` 表示那一分钟没有任何探测记录 */
export type MonitorBucketState = "up" | "down" | "unknown"

/**
 * 最近一小时的可用率状态条。
 *
 * 后端在列表响应里一并返回，前端一次请求就能画完整条图，不用逐行追加请求。
 * 每格对应的时间 = `started_at + index * bucket_seconds`。
 */
export interface MonitorUptimeWindow {
  started_at: string
  bucket_seconds: number
  buckets: MonitorBucketState[]
  /** 窗口内一次探测都没有时为 null——不能显示成 100%，那是撒谎 */
  uptime_rate: number | null
}

/** 监控目标（列表/详情响应，附带最近一次探测结果与可用率状态条） */
export interface MonitorTarget {
  id: number
  cmdb_asset_id: number | null
  ip_address: string
  port: number
  label: string
  check_interval_seconds: number
  is_active: boolean
  created_at: string
  latest_status: MonitorLatestStatus | null
  latest_latency_ms: number | null
  latest_detail: string
  latest_checked_at: string | null
  uptime_window: MonitorUptimeWindow
}

/** 创建监控目标请求 */
export interface MonitorTargetCreate {
  ip_address: string
  port: number
  label?: string
  check_interval_seconds?: number
  is_active?: boolean
  cmdb_asset_id?: number | null
}

/** 单条监控状态变化日志 */
export interface MonitorLogItem {
  id: number
  target_id: number
  label: string
  ip_address: string
  port: number
  status: MonitorLatestStatus
  latency_ms: number | null
  detail: string
  checked_at: string
}

/** 更新监控目标请求 */
export interface MonitorTargetUpdate {
  ip_address?: string
  port?: number
  label?: string
  check_interval_seconds?: number
  is_active?: boolean
  cmdb_asset_id?: number | null
}
